lib/image-stream/parse-partitions.js

```
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2017 resin.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const _ = require('lodash');
+const Bluebird = require('bluebird');
+const MBR = require('mbr');
+const GPT = require('gpt');
+
+/** @const {Number} Maximum number of bytes to read from the stream */
+const MAX_BYTES = 8192;
+
+/** @const {Number} Initial number of bytes read */
+const INITIAL_LENGTH = 0;
+
+/** @const {Number} Initial block size */
+const INITIAL_BLOCK_SIZE = 512;
+
+/** @const {Number} Maximum block size to check for */
+const MAX_BLOCK_SIZE = 4096;
+
+/**
+ * @summary Attempt to parse the GPT from various block sizes
+ * @private
+ * @param {Buffer} buffer - Buffer
+ * @returns {GPT|null}
+ */
+const parseGPT = (buffer) => {
+
+  let blockSize = INITIAL_BLOCK_SIZE;
+  let gpt = null;
+
+  // Attempt to parse the GPT from several offsets,
+  // as the block size of the image may vary (512,1024,2048,4096);
+  // For example, ISOs will usually have a block size of 4096,
+  // but raw images a block size of 512 bytes
+  while (blockSize <= MAX_BLOCK_SIZE) {
+    try {
+      gpt = GPT.parse(buffer.slice(blockSize));
+    } catch (error) {
```

Maybe we worth using `_.attempt` if we want to swallow the error?

```js
while (blockSize <= MAX_BLOCK_SIZE) {
  const gpt = _.attempt(GPT.parse, buffer.slice(blockSize));

  if (!_.isError(gpt)) {
    return gpt;
  }

  blockSize += blockSize;
}
```